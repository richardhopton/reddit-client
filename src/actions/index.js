export const CHOOSE_SUBREDDIT = "CHOOSE_SUBREDDIT";
export const PARSE_POSTS = "PARSE_POSTS";

export const chooseSubreddit = subreddit => ({
  type: CHOOSE_SUBREDDIT,
  subreddit
});

const parsePosts = (subreddit, postsJson) => ({
  type: PARSE_POSTS,
  subreddit,
  posts: postsJson.data.children.map(post => post.data),
  lastUpdated: Date.now()
});

export const fetchPosts = subreddit => dispatch => {
  return fetch(`https://www.reddit.com/r/${subreddit}.json?count=25`)
    .then(response => response.json())
    .then(json => dispatch(parsePosts(subreddit, json)));
};
