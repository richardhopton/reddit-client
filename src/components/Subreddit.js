import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { chooseSubreddit, fetchPosts } from "../actions";
import styles from "./styles.css";

export class Subreddit extends Component {
  componentDidMount() {
    if (!this.props.match) return;
    const { dispatch } = this.props;
    const { subreddit } = this.props.match.params;
    dispatch(chooseSubreddit(subreddit));
  }

  componentDidUpdate(prevProps) {
    const { dispatch, subreddit } = this.props;
    if (prevProps.subreddit === subreddit) return;

    dispatch(fetchPosts(subreddit));
  }

  render() {
    const { subreddit, posts, lastUpdated } = this.props;
    return (
      <div className={styles.subreddit}>
        <div className={styles.header}>
          <div className={styles.headerName}>{subreddit}</div>
          {lastUpdated && (
            <div className={styles.lastUpdated}>
              Last updated: {new Date(lastUpdated).toLocaleTimeString()}
            </div>
          )}
        </div>
        <div className={styles.body}>
          {posts.length === 0 && <h2>Empty.</h2>}
          {posts.length > 0 && posts.map((post,i) => (
            <div key={i}>
              {post.title}
            </div>
          ))}
        </div>
      </div>
    );
  }
}

Subreddit.propTypes = {
  subreddit: PropTypes.string.isRequired,
  posts: PropTypes.array.isRequired,
  lastUpdated: PropTypes.number,
  dispatch: PropTypes.func.isRequired
};

function mapState(state) {
  const { subreddit, posts } = state;
  const { lastUpdated, items } = posts[subreddit] || {
    items: []
  };

  return {
    subreddit,
    posts: items,
    lastUpdated
  };
}

export default connect(mapState)(Subreddit);
