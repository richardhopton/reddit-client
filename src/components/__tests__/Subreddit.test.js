import React from "react";
import { Subreddit } from "../Subreddit";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { CHOOSE_SUBREDDIT } from "../../actions";

const props = {
  subreddit: "oddlysatisfying",
  posts: [{ title: "1" }, { title: "2" }],
  dispatch: () => {}
};

// Check the component hasn't been accidentally adjusted by using a jest snapshot
test("Subreddit renders expected snapshot", () => {
  const tree = renderer.create(<Subreddit {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

// Check interactions
const setup = () => {
  const dispatch = jest.fn();
  const match = { params: { subreddit: "news" } };
  const start = shallow(
    <Subreddit {...props} dispatch={dispatch} match={match} />
  );

  return { dispatch, start };
};

test("Subreddit dispatches choose subreddit action on render", () => {
  const { dispatch, start } = setup();

  const expected = { type: CHOOSE_SUBREDDIT, subreddit: "news" };
  expect(dispatch).toHaveBeenCalledWith(expected);
});
