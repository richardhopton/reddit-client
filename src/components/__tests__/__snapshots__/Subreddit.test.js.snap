// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subreddit renders expected snapshot 1`] = `
<div
  className="subreddit"
>
  <div
    className="header"
  >
    <div
      className="headerName"
    >
      oddlysatisfying
    </div>
  </div>
  <div
    className="body"
  >
    <div>
      1
    </div>
    <div>
      2
    </div>
  </div>
</div>
`;
